### PR TITLE
std.Build: make number of collected stack frames configurable

### PR DESCRIPTION
**Pull request by Sahnvour**
Not sure this is the best way to do it, but I faced a usecase where I needed more stack frames, and `first_ret_addr` is less practical.
Also unsure about the added output, but I think it's best to inform the user that stack traces _could_ be printed via a setting.

The number of stack frames being just a builder variable, it can change during build execution so no assumption that stack traces are present can be made. We could probably make it static in the program via the `@import("root")` mechanism but I don't know how to use it in a build context.

### CODE REVIEW

File: lib/std/Build/Step.zig [hunk 1]
```diff
@@ -157,7 +147,20 @@ pub fn init(options: StepOptions) Step {
         .dependants = .{},
         .state = .precheck_unstarted,
         .max_rss = options.max_rss,
-        .debug_stack_trace = addresses,
+        .debug_stack_trace = blk: {
+            if (options.owner.debug_stack_frames_count > 0) {
```
Comment (andrewrk) [inline lib/std/Build/Step.zig:151]: you don't need this `if`

File: lib/std/Build/Step.zig [hunk 2]
```diff
@@ -157,7 +147,20 @@ pub fn init(options: StepOptions) Step {
...
+                var addresses = arena.alloc(usize, options.owner.debug_stack_frames_count) catch @panic("OOM");
```
Comment (andrewrk) [inline lib/std/Build/Step.zig:152]:
```suggestion
                const addresses = arena.alloc(usize, options.owner.debug_stack_frames_count) catch @panic("OOM");
```

File: lib/std/Build/Step.zig [hunk 3]
```diff
@@ -39,7 +39,7 @@ test_results: TestResults,
 
 /// The return address associated with creation of this step that can be useful
 /// to print along with debugging messages.
-debug_stack_trace: [n_debug_stack_frames]usize,
+debug_stack_trace: ?[]usize,
```
Comment (andrewrk) [inline lib/std/Build/Step.zig:42]: why does it need to be optional now?

File: lib/std/Build/Step.zig [hunk 4]
```diff
@@ -245,11 +250,19 @@ pub fn dump(step: *Step) void {
...
-    };
+    if (step.getStackTrace()) |stack_trace| {
+        w.print("name: '{s}'. creation stack trace:\n", .{step.name}) catch {};
+        std.debug.writeStackTrace(stack_trace, w, ally, debug_info, tty_config) catch |err| {
+            stderr.writer().print("Unable to dump stack trace: {s}\n", .{@errorName(err)}) catch {};
+            return;
+        };
+    } else {
+        const field = "debug_stack_frames_count";
+        comptime assert(@hasField(Build, field));
```
Comment (andrewrk) [inline lib/std/Build/Step.zig:261]: what's going on here?

**Review by andrewrk — CHANGES_REQUESTED** [review]

File: lib/std/Build/Step.zig (hunk 3 above)
Comment (Sahnvour) [reply lib/std/Build/Step.zig:42]: It doesn't, I had this to convey intent. I'll remove it if it's not idiomatic.

File: lib/std/Build/Step.zig (hunk 4 above)
Comment (Sahnvour) [reply lib/std/Build/Step.zig:261]: This is a paranoid attempt at keeping the warning message in sync with the code, checked by the compiler.
I don't mind removing it, I'm just always concerned by documentation/messages silently becoming out of date.

File: lib/std/Build/Step.zig (hunk 3 above)
Comment (andrewrk) [reply lib/std/Build/Step.zig:42]: With it being an optional, it's unclear what's the difference between `null` and being a length 0 stack trace. I think there is no meaningful difference, therefore it should be non-optional.

**Pull request merged**